cupy/cusparse.py

```diff
@@ -0,0 +1,155 @@
+import atexit
+import collections
+
+import six
+
+import cupy
+from cupy import cuda
+from cupy.cuda import cusparse
+
+
+_handles = collections.defaultdict(cusparse.create)
+
+
+def get_handle():
+    dev = cuda.get_device_id()
+    return _handles[dev]
+
+
+@atexit.register
+def reset_handles():
+    global _handles
+    handles = _handles
+    _handles = collections.defaultdict(cusparse.create)
+
+    for handle in six.itervalues(handles):
+        cusparse.destroy(handle)
+
+
+class MatDescriptor(object):
+
+    def __init__(self, descriptor):
+        self.descriptor = descriptor
+
+    @classmethod
+    def create(cls):
+        descr = cusparse.createMatDescr()
+        return MatDescriptor(descr)
+
+    def __del__(self):
+        if self.descriptor:
+            cusparse.destroyMatDescr(self.descriptor)
+            self.descriptor = None
+
+    def set_mat_type(self, typ):
+        cusparse.setMatType(self.descriptor, typ)
+
+    def set_mat_index_base(self, base):
+        cusparse.setMatIndexBase(self.descriptor, base)
+
+
+def _call_cusparse(name, dtype, *args):
+    if dtype == 'f':
+        prefix = 's'
+    elif dtype == 'd':
+        prefix = 'd'
+    else:
+        raise TypeError
+    f = getattr(cusparse, prefix + name)
+    return f(*args)
+
+
+def csr2dense(x, out=None):
+    """Convert CSR-matrix to a dense matrix.
+
+    Args:
+        x (cupy.sparse.csr_matrix): A sparse matrix to convert.
+        out (cupy.ndarray or None): A dense metrix to store the result.
+
+    Returns:
+        cupy.ndarray: Converted result.
+
+    """
+    dtype = x.dtype
+    assert dtype == 'f' or dtype == 'd'
+    if out is None:
+        out = cupy.empty(x.shape, dtype=dtype, order='F')
+    else:
+        out = out.asfortranarray(out)
```

Comment:
Does this line work? It seems that this line contains two problems:

- `cupy.ndarray` does not have `asfortranarray` but `cupy` does.
- `asfortranarray` copies if needed, so the result may not be stored to the original `out`.